Sources/BSON/JSON.swift

```diff
@@ -0,0 +1,139 @@
+import Foundation
+
+/// Enum representing a JSON value, used internally for modeling JSON
+/// during extendedJSON parsing/generation.
+internal enum JSON: Codable {
+    case number(Double)
+    case string(String)
+    case bool(Bool)
+    indirect case array([JSON])
+    indirect case object([String: JSON])
+    case null
+
+    /// Initialize a `JSON` from a decoder.
+    /// Tries to decode into each of the JSON types one by one until one succeeds or
+    /// throws an error indicating that the input is not a valid `JSON` type.
+    public init(from decoder: Decoder) throws {
```

Comment:
`init` and `encode` here can be made `internal` as well.

Reply:
oh okay, I misunderstood our slack conversation from earlier. fixed now